### PR TITLE
fix: edge case where request has invalid multiple values for same field

### PR DESCRIPTION
**Pull request by tungfinblox**
# Reason for This PR

when http request has nested level of data then there will be panic in function `mount` of `dataTree` structure

## Description of Changes

this PR make the `mount` function be able to handle such edge case. Test cases are provided in the PR

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the MIT license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

**Review by paunin — APPROVED** [review]
Should DO!

File: handler/parse.go [hunk 1]
```diff
@@ -38,24 +39,54 @@ func (d dataTree) push(k string, v []string) {
...
+		if !dataTreeOK {
+			oldV := d[i[0]]
+			oldVString, stringOk := oldV.(string)
+			oldVStringArray, stringArrayOk := oldV.([]string)
+			if stringOk && len(oldVString) == 0 {
+				d[i[0]] = make(dataTree)
+			} else if stringArrayOk && len(oldVStringArray) == 0 {
+				d[i[0]] = make(dataTree)
+			} else {
+				panic(
```
Comment (rustatian) [inline handler/parse.go:58]: Hey @tungfinblox 👋🏻 

Thanks for the PR 👍🏻 

RR should not panic. In this case, the user may send invalid data, leading to RR shutdown.

Since we have this edge-case, it's better to return an error instead of panicking and handle it properly (return from the `request` method).

Comment (tungfinblox) [reply handler/parse.go:58]: Got it. Will send update later.

Comment (tungfinblox) [reply handler/parse.go]: I've just updated the PR to not panic as we discussed :+1:

Comment (rustatian) [reply handler/parse.go]: Nice, thanks 👍🏻

**Comment by rustatian** [comment]
@tungfinblox Could you, please add comments on how this new algo works? I understand that there are not many changes, but it would be nice to document them. I'll review the PR early next week. Thanks a lot 👍🏻

File: handler/parse.go [hunk 2]
```diff
@@ -36,74 +37,89 @@ func parseData(r *http.Request) (dataTree, error) {
...
-				v,
-				d,
-			)
+	_, ok := dt[i[0]]
+	if !ok {
+		dt[i[0]] = make(dataTree)
+		return true, nil
+	}
+
+	potentialErr := fmt.Errorf(
```
Comment (rustatian) [inline handler/parse.go:59]: Do not preallocate an error. This is an `unlikely` branch, so you'll usually end up with the costly `fmt` allocation for no reason 😃.

Comment (tungfinblox) [reply handler/parse.go:59]: Got it

File: handler/parse.go [hunk 3]
```diff
@@ -11,56 +13,145 @@ type dataTree map[string]any
...
 	}
+	return nil
+}
+
+func errWhenCreateNode(tree any, keys []string, value any) error {
+	return fmt.Errorf(
+		"invalid value in tree. key: %+v, val: %+v, tree: %+v",
+		keys[0],
+		value,
+		tree,
```
Comment (rustatian) [inline handler/parse.go:53]: The CI fails because you used a `tree` in the error message unescaped (CWE-117). You may have a look here at how to handle this: https://github.com/roadrunner-server/http/blob/master/middleware/log.go#L136.

But actually, I think that we should show the user the whole error message (which can contain sensitive data). We can log it and show a more generic error to the user.

Let's do the following. You can pass the logger from the handler (`h.log`). Then you'll need to unescape the tree using an approach I showed at the bottom.

So, the function `errWhenCreateNode` will be looking smt like this:

Comment (tungfinblox) [reply handler/parse.go]: yeah. I made it so now only the key is included in the error message.

**Comment by rustatian** [comment]
I tried to test this change with our integrational tests (https://github.com/roadrunner-server/rr-e2e-tests), but the output is not the same. I'm not digging at the logic of the changes, but these tests also should be passed.

To check that, you may clone the repo with the e2e tests. Use replace in the `go.mod` to replace http plugin:
```
replace (
	github.com/roadrunner-server/http/v3 => ../plugins/http
)
```
2. Go to the `php_test_files` subfolder and execute `composer u`.
3. Then go to the `plugins->http` and run tests with the `FormData` name.

So, what particularly failed:

Previously, if you tried to add data to the form like:
```go
	form.Add("key", "value")
	form.Add("key", "value2")
```
As a result, only `value2` will be available.
But if you use a `key[]` syntax, both `value` and `value2` will be available under the `key`.

After you change, there is no difference, `key` is the same as `key[]`.

**Comment by tungfinblox** [comment]
> I tried to test this change with our integrational tests (https://github.com/roadrunner-server/rr-e2e-tests), but the output is not the same. I'm not digging at the logic of the changes, but these tests also should be passed.
> 
> To check that, you may clone the repo with the e2e tests. Use replace in the `go.mod` to replace http plugin:
> 
> ```
> replace (
> 	github.com/roadrunner-server/http/v3 => ../plugins/http
> )
> ```
> 
> 2. Go to the `php_test_files` subfolder and execute `composer u`.
> 3. Then go to the `plugins->http` and run tests with the `FormData` name.
> 
> So, what particularly failed:
> 
> Previously, if you tried to add data to the form like:
> 
> ```go
> 	form.Add("key", "value")
> 	form.Add("key", "value2")
> ```
> 
> As a result, only `value2` will be available. But if you use a `key[]` syntax, both `value` and `value2` will be available under the `key`.
> 
> After you change, there is no difference, `key` is the same as `key[]`.

Hey I've just reverted what caused that.

**Comment by rustatian** [comment]
Great, thanks 👍🏻

**Comment by rustatian** [comment]
Hey @tungfinblox 👋🏻 

`rr-e2e-tests` tests are still failing. Could you please update a description with an ample reason for this PR (with an explanation)? I'm sorry, but I still don't understand what case led to a panic (no description) and what you did to resolve that.

**Comment by tungfinblox** [comment]
Hey :wave: 

Sorry that I didn't took care of the old test cases. I've just fixed them now.
About the comment, I tried to put it above the `mount` method but maybe it's not clear enough.
image

Basically what I did is ignore empty value of a key in `dataTree` structure to handle the edge case that we got when using Laravel Backpack. In Go code it's something like this that will cause old code to panic:

```
	form := url.Values{}
	form.Add("arr", "")
	form.Add("arr[x][y][z]", "y")
	form.Add("arr[x][y][e]", "f")
```

I hope that is clearer for you now. Feel free to ask more question so that I can explain more.

**Comment by rustatian** [comment]
Ah, sorry, it should be all `arr`. Now I see 😃

**Review by rustatian — APPROVED** [review]
lgtm

**Comment by rustatian** [comment]
Thanks @tungfinblox 👍🏻
